simperby/src/lib.rs

```diff
@@ -131,12 +197,37 @@ pub trait SimperbyApi {
     /// TODO: define the type of the state.
     async fn get_consensus_status(&self) -> ();
 
-    /// Submits a vote for the given item, identified by its hash.
+    /// Gets the current status of the p2p network.
+    ///
+    /// Unlike the storage, the p2p network operations are done in the background, in other words,
+    /// there is no [`SimperbyApi`] method that directly accesses the p2p network.
+    /// Thus we need a separate API to see the current status of it.
+    ///
+    /// TODO: define the type of the state.
+    async fn get_network_status(&self) -> ();
+
+    /// Gets the `number`-last logs of attempts to execute `SimperbyOperation`s.
+    async fn get_operation_log(&self, number: usize) -> Vec<SimperbyOperationLog>;
+
+    /// Attempts to propose a block for this round. This may update the node state.
+    ///
+    /// It fails
+    /// 1. with the same cause as `check_block`
+    /// 2. if this node is not the current leader.
+    /// 3. if this node has already proposed another block for this round.
+    async fn propose_block(
+        &self,
+        block: Block,
+        round: Round,
+        prevote_signature: TypedSignature<(BlockHeader, Round)>,
+    ) -> Result<(), SimperbyError>;
+
+    /// Submits a vote for the given item, identified by its hash. This may update the node state.
     async fn submit_consensus_vote(
         &self,
         hash: Hash256,
-        signature: TypedSignature<Hash256>,
-    ) -> Result<(), String>;
+        signature: Signature, // This is untyped because the signer doesn't know the source type.
```

Comment:
Aren't signers signing on `ConsensusVoteItem`?

Reply:
No. They sign on `ConsensusVoteItem::hash`.

Reply:
Then what does it mean by `the signer doesn't know the source type.`?
Any problems using TypedSignature<Hash256>?

Reply:
The user doesn't know what type of data was hashed into `hash`.
The Simperby node just instructs where to sign, not what it is.